lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/TaskManager.cs

```diff
@@ -0,0 +1,477 @@
+            UpdateState(completedTask.Id, TaskStateEvent.CompletedTask);
+        }
+
+        /// <summary>
+        /// This method is called when receiving IFailedTask event during task submitting or running
+        /// Removes the task from running tasks if the task was running
+        /// Updates the task state to fail based on the error message in the failed task
+        /// </summary>
+        /// <param name="failedTask"></param>
+        internal void RecordFailedTaskDuringRunningOrSubmissionState(IFailedTask failedTask)
+        {
+            //// Remove the task from running tasks if it exists there
+            _runningTasks.Remove(failedTask.Id);
+            UpdateState(failedTask.Id, GetTaskErrorEvent(failedTask));
+        }
+
+        /// <summary>
+        /// This method is called when receiving IFailedTask event during system shutting down. 
+        /// If the task failed because it receives the close command from driver, update the task state to TaskClosedByDriver.
+        /// Task could fail by communication error or any other application or system error during this time, as long as it is not 
+        /// TaskFailedByEvaluatorFailure, update the task state based on the error received. 
+        /// </summary>
+        /// <param name="failedTask"></param>
+        internal void RecordFailedTaskDuringSystemShuttingDownState(IFailedTask failedTask)
+        {
+            var taskState = GetTaskState(failedTask.Id);
+            if (taskState == StateMachine.TaskState.TaskWaitingForClose)
+            {
+                UpdateState(failedTask.Id, TaskStateEvent.ClosedTask);
+            }
+            else if (taskState != StateMachine.TaskState.TaskFailedByEvaluatorFailure)
+            {
+                UpdateState(failedTask.Id, GetTaskErrorEvent(failedTask));
+            }
+        }
+
+        /// <summary>
+        /// This method is called when receiving an IFailedEvaluator event during TaskSubmitted, TaskRunning or system shutting down.
+        /// Removes the task from RunningTasks if the task associated with the FailedEvaluator is present and running. 
+        /// Sets the task state to TaskFailedByEvaluatorFailure 
+        /// </summary>
+        /// <param name="failedEvaluator"></param>
+        internal void RecordTaskFailWhenReceivingFailedEvaluator(IFailedEvaluator failedEvaluator)
+        {
+            if (failedEvaluator.FailedTask.IsPresent())
+            {
+                var taskId = failedEvaluator.FailedTask.Value.Id;
+                var taskState = GetTaskState(taskId);
+                if (taskState == StateMachine.TaskState.TaskRunning)
+                {
+                    RemoveRunningTask(taskId);
+                }
+
+                if (GetTaskState(taskId) == StateMachine.TaskState.TaskWaitingForClose)
+                {
+                    UpdateState(taskId, TaskStateEvent.ClosedTask);
+                }
+                else
```

Comment:
If the task is in TaskWaitingForClose state, meaning we have sent close event from driver to the task, before the driver received even back from the closing task, we get FailedEvaluator which contains the same task. 
If we simply set the state to FailedTaskEvaluatorError, from failure over perspective, the result is the same because the task ends up to a final state. But if you look at the design document, the transition diagram, TaskWaitingForClose can only be transit to TaskClosedByDriver. I believe that was one of the comments I received in the design review, initially we could point from TaskWaitingForClose to any task failure. I think for TaskWaitingForClose to only transit to TaskClosedByDriver is fine and simple enough, that shows driver wants to close this task. No matter what happens in the process, we would eventually set it as TaskClosedByDriver .